### ISSUE TITLE
[no-unused-vars] False positive when declaring modules

### ISSUE THREAD

**Issue by troch** [typescript-eslint/typescript-eslint#2523]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

I also tried canary versions.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": "error"
  }
}
```

```TS
declare global {
  interface Window {
     /* ... */
  }
}

declare module 'external-module' {
  export default (arg: any) => any
}
```

**Expected Result**

I didn't expect any error.

**Actual Result**

- Error `'global' is defined but never used.`
- Error `'arg' is defined but never used.`

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.1.0` |
| `@typescript-eslint/parser`        | `4.1.0` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.8.1` |
| `node`                             | `12.18.1` |

**Comment by bradzacher** [comment]
Where is your `declare global`?
Is it in a `.d.ts` file?

**Comment by tblue1994** [comment]
We are running into a very similar situation, and yes, it is happening inside a `.d.ts` file

**Comment by bradzacher** [comment]
If you're running the latest version, then there should be no way that it's reporting in a `.d.ts` file.

Could you provide more information please? A comment with no context is counter productive to helping me investigate.

A repro repo is always a great way to help the volunteer maintainers investigate.

**Comment by troch** [comment]
In my case `declare global` is not in a d.ts file but collocated with code.

**Comment by wKovacs64** [comment]
![image](https://user-images.githubusercontent.com/1288694/92546181-307ab200-f20f-11ea-958e-a33b606f04cf.png)

Not sure if this is the same or slightly different - in my case, I'm using declaration merging with a namespace from `node_modules`. Same versions as @troch reported. And I guess it's right, I'm not using `Subject`, but that's the signature of the declaration I'm merging into, so...?

I can try to put together a minimal reproduction soon.

**Comment by zouhangwithsweet** [comment]
![image](https://user-images.githubusercontent.com/21095710/92696422-6db08280-f37c-11ea-9f7b-a6ae2e9e8b53.png)

also in `d.ts` file

**Comment by ldantonov** [comment]
I see the same issue upgrading to v4.1.0 from 4.0.1, in a .ts file:
```
declare global {
  interface HTMLElement {
    _fitSpaceOnResize?: FitSpaceResizeListener
  }
}

```
error  'global' is defined but never used  @typescript-eslint/no-unused-vars

**Comment by davidje13** [comment]
I see the same problem when defining custom Jest matchers:

```typescript
declare global { // <-- 'global' is defined but never used (@typescript-eslint/no-unused-vars)
  namespace jest {
    interface Matchers<R> { // <-- 'Matchers' is defined but never used (@typescript-eslint/no-unused-vars)
      toBeSeven: () => R;
    }
  }
}

expect.extend({
  toBeSeven(value): jest.CustomMatcherResult {
    if (value !== 7) {
      return { pass: false, message: (): string => `Expected seven, but got ${value}` };
    }
    return { pass: true, message: (): string => 'Expected non-seven' };
  },
});

it('is seven', () => {
  expect(7).toBeSeven();
});
```

As others have noted, this is in the main `.ts` file, **not** in a `.d.ts` file. AFAIK this is a pretty standard pattern when defining custom matchers for use within a project.

**Comment by bradzacher** [comment]
@wKovacs64

```ts
declare namespace Cypr_something_I_cant_see {
  interface Chainable<Subject = any> {
    toggleDarkMode(): void;
    clearLocalStorageForReal(): void;
  }
}
```

This case is not a false-positive - the generic is unused.
If the generic is required for some reason, you must use a disable comment.

---

The other cases of using a `declare global` in a `.ts` file will be fixed in the next version

**Comment by davidje13** [comment]
> If the generic is required for some reason, you must use a disable comment.

It isn't required, so the proper fix is to change the code (in this case, just `interface Chainable {` with no generics).

TypeScript will still correctly merge the definitions if one has no generics (or only some of the generics). I confirmed that when I changed `Matchers<R, T>` to just `Matchers<R>` in my Jest example, because the rule is correctly catching unused generic arguments.

---

Thanks for the fix!

**Comment by wKovacs64** [comment]
Ah, learn something new every day - thanks, @davidje13 and @bradzacher!
